src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -46,18 +50,93 @@ protected virtual CompletionItemRules GetCompletionItemRules(IReadOnlyList<ISymb
-                    group symbol by texts into g
-                    select this.CreateItem(
-                        g.Key.displayText, g.Key.suffix, g.Key.insertionText, g.ToList(), context,
-                        invalidProjectMap: null, totalProjects: null, preselect: preselect);
+                var itemListBuilder = ImmutableArray.CreateBuilder<CompletionItem>();
 
-            return q.ToImmutableArray();
+                foreach (var symbolGroup in symbolGroups)
+                {
+                    var item = this.CreateItem(
+                            symbolGroup.Key.displayText, symbolGroup.Key.suffix, symbolGroup.Key.insertionText, symbolGroup.ToList(), context,
+                            invalidProjectMap: null, totalProjects: null, preselect: preselect);
+
+                    if (IsTargetTypeCompletionFilterExperimentEnabled(context.Workspace))
+                    {
+                        if (symbolGroup.Any(s => ShouldIncludeInTargetTypedCompletionList(s, inferredTypes, context.SemanticModel, context.Position)))
+                        {
+                            item = item.AddTag(WellKnownTags.MatchingType);
+                        }
+                    }
+
+                    itemListBuilder.Add(item);
+                }
+
+                return itemListBuilder.ToImmutable();
+            }
+            else
+            {
+                var q = from symbol in symbols
+                        let texts = GetDisplayAndSuffixAndInsertionText(symbol, context)
+                        group symbol by texts into g
+                        select this.CreateItem(
+                            g.Key.displayText, g.Key.suffix, g.Key.insertionText, g.ToList(), context,
+                            invalidProjectMap: null, totalProjects: null, preselect: preselect);
+
+                return q.ToImmutableArray();
+            }
+        }
+
+        private bool IsTargetTypeCompletionFilterExperimentEnabled(Workspace workspace)
+        {
+            if (!_isTargetTypeCompletionFilterExperimentEnabled.HasValue)
+            {
+                var experimentationService = workspace.Services.GetService<IExperimentationService>();
+                _isTargetTypeCompletionFilterExperimentEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.TargetTypedCompletionFilter);
+            }
+
+            return _isTargetTypeCompletionFilterExperimentEnabled == true;
+        }
+
+        private bool ShouldIncludeInTargetTypedCompletionList(ISymbol symbol, ImmutableArray<ITypeSymbol> inferredTypes, SemanticModel semanticModel, int position)
+        {
+            // When searching for identifiers of type C, exclude the symbol for the `C` type itself.
+            if (symbol.Kind == SymbolKind.NamedType)
+            {
+                return false;
+            }
+
+            // Avoid offering object.ToString() when a string is expected, for example.
```

Comment:
why? :)

Reply:
I found it to be noisy when playing with it. Of course, if you have your own custom ToString it will still show up.

Reply:
> I found it to be noisy when playing with it. Of course, if you have your own custom ToString it will still show up.

interesting.  not opposed to this.  just putting on my jason cap.  is there a more general rule to be followed here, or is it really just ToString that is problematic.  will GetHashCode be annoying in an 'int' context for example?

Reply:
I found GetHashCode annoying too, and it's excluded too below. I think the problem is my comment only pointing out ToString?

Reply:
oh fair enough.  yeah, we miht want to simply say that the default System.Object members ('ToString/GetHashCode/Equals') show up too often since they commonly match string/int/bool.  But end up feeling very noisy and not in line with what is commonly wanted when filtering.